sphinxval/utils/duplicates.py

```diff
@@ -0,0 +1,417 @@
+    dict = initialize_forecast_dict()
+
+    #Loop through the forecasts for each model and fill in quantity_dict
+    #as appropriate
+    for ix in range(len(model_objs)):
+        fill_forecast_dict_row(ix, model_objs[ix], dict)
+ 
+    df = pd.DataFrame(dict)
+    
+    return df
+    
+
+
+def remove_forecast_duplicates(all_energy_channels, model_objs):
+    """ Remove any duplicated Forecast objects from the model_objs array.
+    
+    """
+    
+    fcast_df = {}
+
+    for energy_key in all_energy_channels:
+        df = fill_forecast_df(model_objs[energy_key])
+
+        #Check for duplicated forecasts and remove
+        df, dup_indices = identify_forecast_duplicates(df)
+        fcast_df.update({energy_key: df})
+
+        for i in sorted(dup_indices, reverse=True):
+            logger.warning(f"DUPLICATE INPUT FORECAST: Removing duplicated forecast for energy channel {energy_key},  {model_objs[energy_key][i].source}")
+            model_objs[energy_key].pop(i)
+        
+    return model_objs, fcast_df
+
+
+
+def remove_resume_duplicates(r_df, fcast_df, model_objs):
+    """ Compare exact filenames in the resume dataframe with filenames
+        in the dataframe of Forecast objects from model_objs
+    
+        INPUTS:
+        
+            :r_df: (pandas DataFrame) SPHINX_dataframe read back in from a previous run
+            :fcast_df: (dict of DataFrames) df sorted by energy channel containing the unique
+                forecasts read in to sphinx. Created in remove_forecast_duplicates().
+            :model_objs: (dict of Forecast objects) unique forecast objects sorted by energy channel
+            
+        OUTPUTS:
+        
+            :model_objs: (dict of Forecast objects) forecasts with filenames already
+                present in r_df have been removed
+    
+    """
+    
+    for energy_key in model_objs.keys():
+        df = fcast_df[energy_key]
+
+        df_dup = df[df['Forecast Source'].isin(r_df['Forecast Source'])]
+        dup_indices = df_dup['Prediction Index'].to_list()
+        
+        for i in sorted(dup_indices, reverse=True):
+            logger.warning(f"DUPLICATE RESUME FORECAST: Removing duplicated forecast already present in the resume SPHINX_dataframe for energy channel {energy_key}, {model_objs[energy_key][i].source}")
+            model_objs[energy_key].pop(i)
+
+
+    return model_objs
+
+
+def remove_sphinx_duplicates(df):
+    """ Check the SPHINX dataframe for duplicate entries. Issue warning
+        and remove repeated forecasts, combined with observatory information.
+        
+        Forecasts will be considered duplicate if all fields in the
+        dataframe are exactly the same.
+        
+        Output:
+        
+            :df: (dataframe) with unique entries
+        
+    """
+    #Extract key rows from the df that uniquely identify a forecast
+    #Cannot use all df entries, because the hash command cannot hash lists.
```

Comment:
Again, would be better to enumerate what **not** to include instead of this very long list.